pkg/dio/writer.go

```diff
@@ -32,65 +32,51 @@ type DynamicMultiWriter interface {
 }
 
 type multiWriter struct {
-	mutex     sync.Mutex
-	waitGroup sync.WaitGroup
-
+	mutex   sync.Mutex
 	writers []io.Writer
 }
 
 func (t *multiWriter) Write(p []byte) (int, error) {
 	var n int
 	var err error
 
-	var wTmp []io.Writer
-	if verbose {
-		defer func() {
-			log.Debugf("[%p] write %q to %d writers (err: %#+v)", t, string(p[:n]), len(wTmp), err)
-		}()
-	}
-
 	t.mutex.Lock()
-
-	t.waitGroup.Add(1)
-	defer t.waitGroup.Done()
-
-	// stash a local copy of the slice as we never want to write twice to a single writer
-	// if remove is called during this flow
-	wTmp = make([]io.Writer, len(t.writers))
-	copy(wTmp, t.writers)
-
-	t.mutex.Unlock()
+	defer t.mutex.Unlock()
 
 	eof := 0
 	// possibly want to add buffering or parallelize this
-	for _, w := range wTmp {
+	for _, w := range t.writers {
 		n, err = w.Write(p)
```

Comment:
If I recall correctly this call can block depending on the Writer type - this is one of the reasons that we didn't hold the lock while performing the write.

What is the purpose and intent behind removing all of the parallelism from these tests? Generally that's where we would observe the most issues so this seems to be reducing the efficacy of the tests. And in this specific case, of the functional package code as well.